### Introduction
Codes and instructions for comparing SubseqHash2 (SubseqHash2r variant), SubseqHash2w, SubseqHash, minimizer, syncmer and strobemer on readmapping 

### Usage
    
- To generate random tables for SubseqHash/SubseqHash2
  ```
  ../src/gentable.out k d ./table_saving_path
  ../src/gentable_subseqhash1.out k d t ./table_saving_directory
  ```

	 - ` ../src/gentable.out k d ./table_saving_path` generates random tables for SubseqHash2, where `k` is the length of subsequence, `d` is a parameter, and `./table_saving_path` is the path of the file saving the tables. A sample file is at `./table_subseq2/25_11` with k=25 and d = 11.
	- `../src/gentable_subseqhash1.out k d t ./table_saving_directory`generates random tables for SubseqHash, where `k` is the length of subsequence, `d` is a parameter, `t` is the number of repeating, and `./table_saving_directory` is the directory saving the files. There will be `t` files in that directory, and each one has one set of random tables for SubseqHash.  A sample folder is at `./table_subseq1/25_11` with k = 25, d = 11 and 10 files with different tables in it.


- To get the results:
  - Run `../src/alignment_sub2.out data_file n k d t table_file` to get the sequence alignment results using SubseqHash2. `n`, `k` and `d` are the parameters, `t` is the number of repeating times. The table file must be generated with the same parameters. The data file must be generated by the sequence simulation. An example is `../src/alignment_sub2.out ./data/sample 30 25 11 2  ./table_subseq2/25_11`.
  - Run `../src/alignment_substro.out data_file n k d t k0 table_file` to get the sequence alignment results using SubseqHash2w. `k0` is the length of substring concatenating to the subsequence. An example is `../src/alignment_substro.out ./data/sample 30 25 11 2 5 ./table_subseq2/25_11`.
  - Run `../src/alignment_sub1.out data_file n k d t table_directory` to get the sequence alignment results using SubseqHash. An example is `../src/alignment_sub1.out ./data/sample 30 25 11 2 ./table_subseq1/25_11`.
  - Run `../src/alignment_mini.out data_file k w` to get the sequence alignment results using minimizer. `k` is the length of kmer, `w` means to only select the minimum kmer over the w consecutive kmers. An example is `../src/alignment_mini.out ./data/sample 30 25`.
  - Run `../src/alignment_sync.out data_file k s pos1 pos2 .. poss` to get the sequence alignment results using syncmer. `k` is the length of syncmer, `s` and `pos1` to `poss` are the parameters for syncmer. For closed syncmer, the positions are always 0 and k - s. An example is `../src/alignment_sync.out ./data/sample 25 10 0 15`.
  - Run `../src/alignment_stro.out data_file k w w_min w_max t` to get the sequence alignment results using strobemer. `k` is the length of substring, `w` is the number of substrsings to be concatenated,  `w_min` and  `w_max` are the positions for the window, `t` is the number of repeating times. An example is `../src/alignment_stro.out ./data/sample 10 2 10 20 2`.
  - The output format is parameters, number of seed-matches, number of true seed-matches, the average ratio of true seed-matches, coverage of true seed-matches, coverage of false seed-matches, matching coverage of true seed-matches, the island size, the density of seeds, the running time of seeding, and the runinng time of seed matching.